File: Python/Code/boosting.py
#! /usr/bin/env python

import math, numpy, scipy, sys, random
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.svm import SVC
from sklearn import cross_validation

from lib import util, config

class BoostingSVM(object):

    ''' Boosting - using different (sample weighted) Support Vector Machines as underlying models '''

    def __init__(self, n_models):
        self.n_models   = n_models
        self.clf        = list()
        self.w          = list()
        self.alpha      = numpy.matrix(numpy.zeros((self.n_models, 1)))
        self.eps        = numpy.matrix(numpy.zeros((self.n_models, 1)))

    def get_classifier(self):
        return SVC(C = 1, kernel = 'linear', class_weight = 'auto')

    def fit(self, x, y):
        '''fit the training data to all the classifiers'''
        n_samples, n_features = x.get_shape()
        self.w.append((1.0 / n_samples) * numpy.matrix(numpy.ones(n_samples)))
        for i in xrange(0, self.n_models):
            clf = self.get_classifier()
            clf.fit(x, y, sample_weight = numpy.array(self.w[-1])[0])
            I = numpy.matrix(map(lambda f: int(f), clf.predict(x) != y))
            self.eps[i] = (self.w[-1] * I.transpose()) / self.w[-1].sum(1)
            self.alpha[i] = math.log((1 - self.eps[i]) / self.eps[i])
            self.w.append(numpy.multiply(self.w[-1], numpy.exp(self.alpha[i] * I)))
            self.clf.append(clf)

    def score(self, x, y):
        '''return the accuracy of prediction on testing data'''
        predictions = None
        for i in xrange(0, self.n_models):
            if predictions is None:
                predictions = self.clf[i].predict(x)
            else:
                predictions = numpy.vstack((predictions, self.clf[i].predict(x)))
        predictions = numpy.sign(predictions.transpose() * self.alpha)
        return numpy.mean(predictions == y)

def main():
    # initialize global data
    vec = TfidfVectorizer(ngram_range = (1, 5), strip_accents = None, charset_error = 'ignore', stop_words = None)
    labels, stories = util.get_distress_data(config.CONNECTION)
    instances = vec.fit_transform(stories); labels = numpy.array(labels);
    random.seed(0)

    n_models = 5; cv = 5; cv_accuracy = list();
    for i in xrange(0, cv):
        print "Iteration #" + str(i) + "..."

        # initialize training/testing data
        cv_data = cross_validation.train_test_split(instances, labels, test_size = 0.5, random_state = i)
        x_training = cv_data[0]
        x_testing = cv_data[1]
        y_training = cv_data[2]
        y_testing = cv_data[3]

        # initialize the classifier
        clf = BoostingSVM(n_models)
        clf.fit(x_training, y_training)

        # measure prediction accuracy
        cv_accuracy.append(clf.score(x_testing, y_testing))

    print "Scores => " + str(cv_accuracy)
    print "Mean   => " + str(numpy.mean(cv_accuracy))

if __name__ == "__main__":
    main()

File: MATLAB/Code/boosting.m
clear; close all;

load('seeds.mat'); rng(s);

[labels, instances] = libsvmread('Data/n-gram.data');

M = 9;
cv = cvpartition(labels, 'Kfold', 10);
cv_accuracy = zeros(1, cv.NumTestSets);
param = '-t 0 -c 1 -h 0 -w1 %.3f -w-1 %.3f';

for i = 1 : cv.NumTestSets
    training = cv.training(i);
    testing = cv.test(i);
    
    x_training = instances(training, :); y_training = labels(training, :);
    x_testing = instances(testing, :); y_testing = labels(testing, :);
    
    models = cell(M, 1);
    n = size(x_training, 1);
    w = repmat(1 / n, n, M);
    alpha = zeros(M, 1);
    eps = zeros(M, 1);
    
    for m = 1 : M
        positive = size(y_training, 1) / sum(y_training == 1);
        negative = size(y_training, 1) / sum(y_training == -1);
        models{m} = svmtrain(w(:, m) ./ min(w(:, m)), y_training, x_training, sprintf(param, positive, negative));

        predictions = svmpredict(y_training, x_training, models{m});

        I = (predictions ~= y_training);

        eps(m) = (w(:, m)' * I) / sum(w(:, m));
        alpha(m) = log ( (1 - eps(m)) / eps(m) );

        if m < M
            w(:, m + 1) = w(:, m) .* exp(alpha(m) * I);
        end
    end
    
    predictions = zeros(size(y_testing, 1), M);
    for m = 1 : M
        predictions(:, m) = svmpredict(y_testing, x_testing, models{m});
    end
    predictions = sign(predictions * alpha);
    
    cv_accuracy(i) = sum(predictions == y_testing) / size(y_testing, 1);
end

fprintf('Accuracy => [%s]\nMean => %s\n', num2str(cv_accuracy), num2str(mean(cv_accuracy) * 100));
